### DISCUSSION TITLE
Merging grains issue

### DISCUSSION THREAD

**Discussion by NikitaSakharov** [mtex-toolbox/mtex#1041] [Ask Anything] [answered]
Hi everyone!

I have some difficulties with grains merging. I hope someone will help.  So thanks in advance!

My EBSD map contains both deformed and recrystallized grains and what I'm trying to do is to merge recrystallized grains only. And I get an error.

Here is my code after I've imported ebsd data & stored them in 'ebsd' variable.
```matlab
%now reconstruct grains
[grains_initial,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',[2*degree, 15*degree]);
%delete all of them less then 5 pixels
ebsd(grains_initial(grains_initial.grainSize<5)) = [];
%and now re-reconstruct grains using corrected ebsd data
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',[2*degree, 15*degree]);
%smooth them
smooth_const = 5;
grains = smooth(grains, smooth_const);
%find recrystallized and deformed grains using GOS criteria
grains_recr = grains(grains.GOS ./ degree < 5);
grains_def = grains(grains.GOS ./ degree >= 5);
%now define twinning direction as one which is rotated 60 deg in respect to (111)
%define crystal symmetry first
CS = ebsd('Copper').CS;
%define rotation direction
m = Miller(1, 1, 1, CS);
%and define rotation axis
rot = rotation.byAxisAngle(m, 60 * degree);
twinning = rot;
isTwinning = angle(gB.misorientation, twinning) < 5*degree;

%And now when I try to merge recrystallized grains with the following command
[mergedGrains,parentId] = merge(grains_recr,twinBoundary,'calcMeanOrientation');
```
I get an error
```
Array indices must be positive integers or logical values.

Error in grain2d/merge (line 191)
grainsMerged.poly(newInd) = grains.poly(keepInd);
```

**Comment by kilir** [comment, marked as answer]
Hi Nikita,

I can't reproduce your observation
```matlab
mtexdata copper

[grains_initial,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',[2*degree, 15*degree]);
%delete all of them less then 5 pixels
ebsd(grains_initial(grains_initial.grainSize<5)) = [];
%and now re-reconstruct grains using corrected ebsd data
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',[2*degree, 15*degree]);

grains = smooth(grains, 5);
%find recrystallized and deformed grains using GOS criteria
grains_recr = grains(grains.GOS <  3*degree);
grains_def = grains(grains.GOS  >= 3*degree);

%define CSL3 twins
CS = ebsd('Copper').CS;
%define rotation direction
m = Miller(1, 1, 1, CS);
%and define rotation axis
twinning = orientation.byAxisAngle(m, 60 * degree);

gB = grains.boundary('C','C')

isTwinning = angle(gB.misorientation, twinning) < 5*degree;

[mergedGrains,parentId] = merge(grains_recr,gB(isTwinning));
```

Does the above example work for you?

Cheers,
Rüdiger

**Reply by NikitaSakharov** [reply]
Hello, Rudiger!

This example does not work for me. 
I get an error

Error using sparse
Index exceeds array bounds.

Error in grain2d/merge (line 68)
    A = A | sparse(mergeId(:,1),mergeId(:,2),1,maxId+1,maxId+1);

**Reply by NikitaSakharov** [reply]
It's interesting. 

Now I get following error message

Array indices must be positive integers or logical values.

Error in grain2d/merge (line 191)
grainsMerged.poly(newInd) = grains.poly(keepInd);

**Reply by NikitaSakharov** [reply]
Maybe it is smthng wrong with my raw ebsd data? Should I provide you with it?

**Reply by NikitaSakharov** [reply]
And now it's saying something wrong with line 199 of merge.m file

Array indices must be positive integers or logical values.

Error in grain2d/merge (line 199)
grainsMerged.poly(newInd) = grains.poly(keepInd);

**Reply by NikitaSakharov** [reply]
And one interesting thing is that code works nearly fine on ebsd data from 'mtexdata copper'.
It merges grains without errors. But it plots them in a very strange way after I added option 'calcMeanOrientation' to merge command (look at the attached picture).

So, at least it merges the grains without error.

So what can be wrong with my ebsd data? Maybe it is too large (600x400 pixels with many small grains)?
![untitled](https://user-images.githubusercontent.com/62562246/119621438-ca6ea300-be0e-11eb-8425-8c1c841a94d3.jpg)

**Reply by kilir** [reply]
Hi Nikita,
that is indeed a very interesting plot!
If you like, you can send me the data, I can try and have a look.
Cheers,
Rüdiger

**Reply by NikitaSakharov** [reply]
Rudiger, many thanks for assistance!

I've attached cpr & crc files in zip-archive
[M0b_x500_220.zip](https://github.com/mtex-toolbox/mtex/files/6551402/M0b_x500_220.zip)
.

**Reply by kilir** [reply]
Hi Nikita,
your issue originates from your intent to merge a subset of the grains without recalculating grains. Using several conditions on the entire dataset might be a better option, e.g. boundary is a twinboundary AND belong to a certain type of grain.

Here's an example which works with your data for me:
```
%% some preparation
[grains_initial,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',[2*degree, 15*degree]);
%delete all of them less then 5 pixels
ebsd(grains_initial(grains_initial.grainSize<5)) = [];
%and now re-reconstruct grains using corrected ebsd data
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',[2*degree, 15*degree]);
grains = smooth(grains, 5);
%%
%find ids ofrecrystallized and deformed grains using GOS criteria
rxx_id = grains(grains.GOS <  3*degree).id;
def_id = grains(grains.GOS >=  3*degree).id;

%define twinning
twinning = orientation.byAxisAngle(Miller(1, 1, 1, ebsd('Copper').CS), 60 * degree);

gB = grains.boundary('C','C')

isTwinning = angle(gB.misorientation, twinning) < 5*degree;

% twinboundaries
twinb = gB(isTwinning)

% twinboundary which belong to to some extent to rxx grains
% both sides
twinboundary_rxx2rxx = twinb(all(ismember(twinb.grainId,rxx_id),2))
% just one side
twinboundary_rxx2all = twinb(any(ismember(twinb.grainId,rxx_id),2))
%%
plot(grains('id',rxx_id),'displayname','rxx','faceColor',[0.94 0.85 0.85])
hold on
plot(grains('id',def_id),'displayname','def','faceColor',[0.75 0.75 0.95])
hold on
plot(grains.boundary)
hold on
plot(twinboundary_rxx2all,'linecolor','b','linewidth',3,'displayname','tb rxx2all')
hold on
plot(twinboundary_rxx2rxx,'linecolor','r','linewidth',1,'displayname','tb rxx2rxx')
hold off
%% merge 
% merge only rxx2rxx
[mergedGrainsrxx2rxx] = merge(grains,twinboundary_rxx2rxx);
% merge rxx2all other grains
[mergedGrainsrxx2all] = merge(grains,twinboundary_rxx2all);

figure
plot(mergedGrainsrxx2rxx,'displayname','merged rxx2rxx')
nextAxis
plot(mergedGrainsrxx2all,'displayname','merged rxx2all')
```

`'calcMeanOrientation'` works as well, however I still think using the mean is not a good idea.

Cheers,
Rüdiger

**Comment by ralfHielscher** [comment]
Hi Nikita,

this is indeed a bug in MTEX related to the situation that you want to apply merge to some subset of grains. I have fixed issue in 2ddc50b. Are you able to apply the changes and test them?

Ralf.

**Reply by NikitaSakharov** [reply]
Hello, Ralf.

I have finally found a minute to work with merging issue.
Unfortunately modified merge.m file does not work too. 
I get following error ('merge2' is the name of modified file):

Undefined function 'calcPolygons' for input arguments of type 'double'.

Error in merge2 (line 202)
[grainsMerged.poly(newInd), grainsMerged.inclusionId(newInd)] = ...

And by the way, Rudiger's workaround seems to work fine. 
I have some questions regarding interpretation of results from physical (or material science) point of view, but it works without error at least.

Rudiger, Ralf, thanks a lot!

**Reply by kilir** [reply]
Hi Nikita,
it seems you are not using the latest version with the fix (the line you get the error message from is from an older version of the function). See here: https://github.com/mtex-toolbox/mtex/blob/develop/EBSDAnalysis/%40grain2d/merge.m

Cheers,
Rüdiger

**Reply by NikitaSakharov** [reply]
Helo,  Rudiger!

It seems not to work anyway. 

I've taken the file from the link you provided (https://github.com/mtex-toolbox/mtex/blob/develop/EBSDAnalysis/%40grain2d/merge.m ), saved it as merge2.m and ran the  code you suggested earlier. 
And ge the same error unfortunately:

[grains_initial,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',[2*degree, 15*degree]);
%delete all of them less then 5 pixels
ebsd(grains_initial(grains_initial.grainSize<5)) = [];
%and now re-reconstruct grains using corrected ebsd data
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',[2*degree, 15*degree]);
grains = smooth(grains, 5);
%find recrystallized and deformed grains using GOS criteria
grains_recr = grains(grains.GOS <  3*degree);
grains_def = grains(grains.GOS  >= 3*degree);
%define CSL3 twins
CS = ebsd('Copper').CS;
%define rotation direction
m = Miller(1, 1, 1, CS);
%and define rotation axis
twinning = orientation.byAxisAngle(m, 60 * degree);
gB = grains.boundary('C','C')
isTwinning = angle(gB.misorientation, twinning) < 5*degree;
[mergedGrains,parentId] = merge2(grains_recr,gB(isTwinning));
 
gB = grainBoundary (show methods, plot)
 
 Segments  mineral 1  mineral 2
    37356     Copper     Copper
Index exceeds array bounds.

Error in merge2 (line 195)
grainsMerged.boundary.grainId(ind) = old2newId(grains.boundary.grainId(ind));

**Reply by ralfHielscher** [reply]
Hi Nikita,

have you placed the function merge2 into the same folder as merge ?

Ralf.

**Reply by NikitaSakharov** [reply]
No, I've added the folder where merge2.m is located to  Matlab Search path

**Reply by ralfHielscher** [reply]
This explains the error you got :) the function needs to be in the same folder.

**Reply by NikitaSakharov** [reply]
Ralf, many thanks! It works now!

**Comment by NikitaSakharov** [comment]
Hello Rudiger and Ralf!
 
Many many thanks for help!
 
I’ll surely test it. I hope I’ll be able to do it tomorrow if Ihave time.
 
Sincerely, Nikita
  
>Четверг, 27 мая 2021, 22:49 +03:00 от Ralf Hielscher ***@***.***>:
> 
> 
>Hi Nikita,
>this is indeed a bug in MTEX related to the situation that you want to apply merge to some subset of grains. I have fixed issue in  2ddc50b . Are you able to apply the changes and test them?
>Ralf.
>—
>You are receiving this because you authored the thread.
>Reply to this email directly,  view it on GitHub , or  unsubscribe . 
 
 
--
Никита Сахаров
